Selector: Make empty attribute selectors work in IE again

mgol opened the pull request:
### Summary ###
<!--
Describe what this PR does. All but trivial changes (e.g. typos)
should start with an issue. Mention the issue number here.
-->

qSA in IE often (but not always) doesn't find elements with an empty name
attribute selector (`[name=""]`). ~~Assigning a temporary name attribute to
the document root (& removing it afterwards) seems to resolve the issue.~~
EDIT: Detect that & fall back to Sizzle traversal.

Fixes gh-4435

### Checklist ###
<!--
Mark an `[x]` for completed items, if you're not sure leave them unchecked and we can assist.
-->

* [x] All authors have signed the CLA at https://cla.js.foundation/jquery/jquery
* [x] New tests have been added to show the fix or feature works
* [x] Grunt build and unit tests pass locally with these changes
* ~~If needed, a docs issue/PR was created at https://github.com/jquery/api.jquery.com~~

<!--
Thanks! Bots and humans will be around shortly to check it out.
-->

mgol commented:
+21 bytes

This only fixes the issue in IE, not Edge. See the comment at https://github.com/jquery/jquery/issues/4435#issuecomment-541442015 for the rationale.

mgol commented:
We could also accept the perf hit & assign such a temporary name attribute in all browsers, at least once. Hopefully that wouldn't cause style recalculation as the final value would be the same as the initial one, worth checking.

mgol commented on src/selector.js:
```diff
@@ -436,6 +437,15 @@ function setDocument( node ) {
+	// Adding a temporary attribute to the document before the selection works
+	// around the issue.
+	if ( isIE && documentElement ) {
+		documentElement.setAttribute( "name", "" );
```
I guess I should check if there's such an attribute already set and - if it is - restore it afterwards. `documentElement` with a `name` attribute would be quite a weird thing to do but it's not impossible.

timmywil replied:
I wonder if the problem is already resolved if the `documentElement` already has a `name` attribute.

mgol replied:
@timmywil It looks like the problem doesn't exist if an _empty_ name attribute is already present on the document. However, a non-empty attribute doesn't help so this is all needed.

mgol commented:
PR updated.

I build a test case similar to https://jsbin.com/cocaqis/1/edit?html,js but with one iframe being available immediately in the DOM (with similar DOM contents) and another similar one being added after a few seconds. If there was no workaround, all of them failed. If I added a workaround in the main doc, all succeeded. If I added one in the first iframe, the selection in the main DOM failed & then the selection in both iframes succeeded. If I added it to the second one, it was failing until the selection happened in that second iframe, then all started suceeding.

Also, if I see success, after refreshing the page (with "Always refresh from server" checked in the Network DevTools tab) I see every selection succeeding as well so the workaround even survives page refreshes.

Of course, IE being IE, it's not impossible that the workaround stops working after a while but for now I think we can work on the assumption it survives long enough. If that proves not enough, we'll depend on someone submitting an issue.

mgol commented:
+30 bytes, BTW.

mgol commented:
Sizzle version of the PR: https://github.com/jquery/sizzle/pull/461

mgol commented:
PR updated. `[name='']` now falls back to Sizzle traversal & both IE 11 & Edge 15-18 pass all tests. +49 bytes.

gibson042 commented on src/selector/rbuggyQSA.js:
```diff
@@ -23,6 +25,18 @@ if ( isIE ) {
+input.setAttribute( "name", "" );
+testEl.appendChild( input );
+if ( !testEl.querySelectorAll( "[name='']" ).length ) {
+	rbuggyQSA.push( "\\[" + whitespace + "*name" + whitespace + "*=" +
```
Does the `=` here also need to cover other operators (e.g., `[*^$]?=`)? Cf. checks for [empty strings after ^= or $= or *=](https://github.com/jquery/sizzle/blob/a528ae520245b13a6a879041e4a2b135c7d6e522/src/sizzle.js#L816) and [case-sensitivity of name attribute](https://github.com/jquery/sizzle/blob/a528ae520245b13a6a879041e4a2b135c7d6e522/src/sizzle.js#L858).

mgol replied:
Thanks, I'll have a look. Unfortunately, I merged the Sizzle PR before seeing this comment so it might need a followup PR later.

mgol replied:
@gibson042 What's the meaning of `!` inside of a regex there? Native CSS selectors don't support negated attribute selector `[name!=value]` so they'd go the Sizzle route anyway.

gibson042 replied:
> @gibson042 What's the meaning of `!` inside of a regex there? Native CSS selectors don't support negated attribute selector `[name!=value]` so they'd go the Sizzle route anyway.

Looks like it's there to make a longer repeat of https://github.com/jquery/sizzle/blob/a528ae520245b13a6a879041e4a2b135c7d6e522/src/sizzle.js#L87 for gzip.

mgol replied:
@gibson042 I checked the selectors you mentioned, the issue is all of them (`*=`, `^=`, `$=`) require the matched string to be non-empty to return any result. Therefore, all browsers return no results for them so there's nothing to workaround for IE here.

gibson042 commented on src/selector/var/whitespace.js:
```diff
@@ -0,0 +1,6 @@
+define( function() {
+	"use strict";
+
+	// http://www.w3.org/TR/css3-selectors/#whitespace
```
Might as well make this HTTPS now.

mgol replied:
Nice catch. I fixed the URL in the other PR but since this declaration has moved I could have missed that URL.
